kernel/arch/dreamcast/hardware/maple/maple_irq.c

```diff
@@ -113,31 +128,41 @@ static void vbl_autodet_callback(maple_state_t *state, maple_frame_t *frm) {
                 vbl_chk_disconnect(state, p, u);
             }
 
-            state->ports[p].units[0].dev_mask = 0;
+            if(dev)
+                dev->dev_mask = 0;
+
+            state->scan_ready_mask |= 1 << p;
         }
         else {
             /* Not a top-level device -- only detach this device */
             vbl_chk_disconnect(state, p, u);
         }
+
+        maple_frame_unlock(frm);
     }
     else if(resp->response == MAPLE_RESPONSE_DEVINFO) {
         /* Device is present, check for connections */
-        if(!state->ports[p].units[u].valid) {
+        if(!dev) {
 #if MAPLE_IRQ_DEBUG
             dbglog(DBG_KDEBUG, "maple: attach on device %c%c\n",
                    'A' + p, '0' + u);
 #endif
 
+            if (irq_inside_int() && !malloc_irq_safe()) {
+                /* We can't create a device now. Fail silently as the device
+                 * will be re-probed in the next loop of the periodic IRQ. */
+                maple_frame_unlock(frm);
+                return;
+            }
```

Comment:
This check, or something similar, should also be needed to protect the `vbl_chk_disconnect` in the if above it (and where it's called from chk_subdevs. That is what automatically calls `maple_driver_detach` which does the `free()` to the attach's `calloc`.

Reply:
I can move this block at the top of the `vbl_autodet_callback()` function.

Reply:
Yea, I guess that makes sense. I was thinking it to be narrowly applied, but the only thing that can be done here if it fires in the middle of a malloc is to update it's function data which is kind of pointless. And either way that check is just to protect against an edge case so hopefully it wouldn't cause significant issue. 

Fortunately that would just mean waiting 4 frames until the autodetect fires again, and is only for device attach/remove so it shouldn't be much of an issue at all.

Reply:
Done.